### PR TITLE
fix: selected network when there are two with the same chain id. 

### LINKED ISSUE #25763: [Bug]: Network name is showing incorrectly after adding Tenderly as custom network
### Describe the bug

After adding Tenderly as a custom network the name shown is still Ethereum, most likely because ChainID for both network is 1.  While this might seems as a minor issue, in the swap team we often use Tenderly for functionality testing and that allows us to save money. So making sure Tenderly works a custom network is important.

I have investigated the problem and it started with this commit
https://github.com/MetaMask/metamask-extension/commit/35950523128fc06d76d8d385120a7d3cb214ce2a 

The problem doesn't happen with this previous commit
https://github.com/MetaMask/metamask-extension/commit/85cf4306af7cd050fe877400074b2a9f7af9c361

### Expected behavior

_No response_

### Screenshots/Recordings

Before the changes

https://github.com/MetaMask/metamask-extension/assets/6082863/ed56786b-1d27-4007-bbca-fb4e08ec0205

After

https://github.com/MetaMask/metamask-extension/assets/6082863/232ba914-f96e-4fe7-86f8-8cf0690f111e

### Steps to reproduce

1.  Add Tenderly as a custom newtork with the following settings
name: 'Tenderly',
url: 'https://rpc.tenderly.co/fork/cdbcd795-097d-4624-aa16-680374d89a43',
chainID: '1',
symbol: 'ETH'

2. Notice that the network name still says `Ethereum`

### Error messages or log output

_No response_

### Version

develop

### Build type

None

### Browser

Chrome

### Operating system

MacOS

### Hardware wallet

_No response_

### Additional context

_No response_

### Severity

_No response_

### PR DESCRIPTION
**Pull request by montelaidev**
## **Description**

This PR fixes the issue when there are two networks with the same chain id and the default network is selected. 

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/25763

## **Manual testing steps**

1. Add a custom network with the chain id of `1`
2. Switch to this network
3. See that it is selected. 

### **Before**

Screenshot 2024-07-12 at 22 47 14

### **After**

Screenshot 2024-07-12 at 22 50 50

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/components/multichain/app-header/app-header.test.js [hunk 1]
```diff
@@ -175,4 +185,49 @@ describe('App Header', () => {
...
+        type: 'rpc',
+        ticker: 'ETH',
+        chainId: '0x1',
+        rpcUrl: 'https://localhost:8545',
+        nickname: 'Localhost',
+      };
+
+      const mockNetworkConfigurations = {
+        [mockProviderConfig.id]: mockProviderConfig,
+      };
```
[the remark this answers is not in the document]
Comment (montelaidev) [reply ui/components/multichain/app-header/app-header.test.js:202]: Whats the suggestion here?

File: ui/selectors/multichain.ts [hunk 2]
```diff
@@ -42,15 +49,19 @@ export type BalancesState = {
...
+};
+
+// TODO: Remove after updating to @metamask/network-controller 20.0.0
+export type NetworkConfigurationWithId = NetworkConfiguration & { id: string };
+
 export type MultichainNetwork = {
   nickname: string;
   isEvmNetwork: boolean;
   chainId: CaipChainId;
   network: // TODO: Maybe updates ProviderConfig to add rpcPrefs.imageUrl field
```
[the remark this answers is not in the document]
Comment (montelaidev) [reply ui/selectors/multichain.ts:63]: Agreed, it will also be deprecated in the next major

**Comment by montelaidev** [comment]
> Hey @montelaidev, did you specifically ask for @MetaMask/design-system-engineers review for this PR? Trying to figure out why we got tagged as code owners here. Happy for our team to review if so

Hey @georgewrmarshall, there was a change in a design system component that was reverted. I think its ok now, we can remove @MetaMask/design-system-engineers from the reviewers. Thanks for checking in!

File: ui/components/multichain/app-header/app-header.test.js [hunk 3]
```diff
@@ -175,4 +185,48 @@ describe('App Header', () => {
...
```
Comment (ccharly) [inline ui/components/multichain/app-header/app-header.test.js:202]: Nit: actually, the suggestion was not applied 🙈 this one is mainly to group mock declarations altogether (like in "shows rpc url as nickname if there isn't a nickname set")

```suggestion
      const mockNetworkConfigurations = {
        [mockProviderConfig.id]: mockProviderConfig,
      };

```

File: ui/components/multichain/app-header/app-header.test.js [hunk 4]
```diff
@@ -175,4 +185,48 @@ describe('App Header', () => {
...
+  describe('network picker', () => {
+    it('shows custom rpc if it has the same chainId as a default network', () => {
+      const mockProviderConfig = {
+        id: 'custom-rpc-localhost',
+        type: 'rpc',
+        ticker: 'ETH',
+        chainId: '0x1',
+        rpcUrl: 'https://localhost:8545',
+        nickname: 'Localhost',
+      };
```
Comment (ccharly) [inline ui/components/multichain/app-header/app-header.test.js:198]: Also, this one is being used twice, maybe we could move it to the top-level? We could also rename it `mockCustomProviderConfig` WDYT?

File: ui/selectors/multichain.ts [hunk 5]
```diff
@@ -111,17 +123,55 @@ export function getMultichainNetwork(
...
+    // TODO: Update to use network configurations when @metamask/network-controller is updated to 20.0.0
+    // ProviderConfig will be deprecated to use NetworkConfigurations
+    // When a user updates a network name its only updated in the NetworkConfigurations.
+    const evmNetwork: ProviderConfigWithImageUrlAndExplorerUrl =
+      getProviderConfig(state);
+    // Could be undefined for default configurations.
+    const evmNetworkConfigurations = getNetworkConfigurations(state);
+    const evmNetworkConfiguration =
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      evmNetworkConfigurations?.[evmNetwork.id!];
```
Comment (ccharly) [inline ui/selectors/multichain.ts:139]: Could we move this into this code branch:

```ts
if (evmNetwork.type === NETWORK_TYPES.RPC) {
  // Here
  
  ...
}
```

Also, I guess that custom networks always requires an `id` to be defined? If yes, we can also add a custom to justify the "no-non-null-assertion" in this case, WDYT?

File: ui/components/multichain/app-header/app-header.test.js (hunk 4 above)
Comment (montelaidev) [reply ui/components/multichain/app-header/app-header.test.js:198]: There is a small difference between the two providers configs

Comment (ccharly) [reply ui/components/multichain/app-header/app-header.test.js:198]: that's my bad then 🙈

File: ui/selectors/multichain.ts [hunk 6]
```diff
@@ -111,17 +123,55 @@ export function getMultichainNetwork(
...
+        imageUrl: CHAIN_ID_TO_NETWORK_IMAGE_URL_MAP[evmChainIdKey],
+      };
+    }
+
+    let nickname;
+    if (evmNetwork.type === NETWORK_TYPES.RPC) {
+      // Could be undefined for default configurations.
+      const evmNetworkConfigurations = getNetworkConfigurations(state);
+      const evmNetworkConfiguration =
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (ccharly) [inline ui/selectors/multichain.ts:153]: I'm not sure if this is true though? It's definitely true for the `rpcUrl` but IDK about `id`, do you know?

```suggestion
        // id will always be defined for custom networks.
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```

File: ui/selectors/multichain.ts [hunk 7]
```diff
@@ -111,17 +123,55 @@ export function getMultichainNetwork(
...
+        evmNetworkConfigurations?.[evmNetwork.id!];
+
+      // These are custom networks defined by the user.
+      // If there aren't any nicknames, the RPC URL is displayed.
```
Comment (ccharly) [inline ui/selectors/multichain.ts:157]: I'd move this at the beginning of the scope (just after the `if (...) {`, to match the `else` pattern

**Review by ccharly — APPROVED** [review]

**Review by matthewwalsh0 — APPROVED** [review]

**Comment by darkwing** [comment]
Do we know how/when this regressed?  Please test re: v12, 12.1, etc, because this is problematic

**Comment by montelaidev** [comment]
> Do we know how/when this regressed? Please test re: v12, 12.1, etc, because this is problematic

 It looks like it is in v12.1 from this https://github.com/MetaMask/metamask-extension/commit/35950523128fc06d76d8d385120a7d3cb214ce2a commit when it started using the multichain networks from this [b3c810e](https://github.com/MetaMask/metamask-extension/commit/b3c810e18ff79cd41e2a15a868ddd703d5a00780) commit

**Pull request merged**